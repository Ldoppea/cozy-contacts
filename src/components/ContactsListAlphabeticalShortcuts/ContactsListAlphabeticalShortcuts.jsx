import React, { useContext } from 'react'
import PropTypes from 'prop-types'

import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import SearchContext from '../Contexts/Search'
import { categorizeContacts } from '../../helpers/contactList'

const ContactsListAlphabeticalShortcuts = ({ contacts }) => {
  const { t } = useI18n()
  
  const isCategorisedList = () => {
    const { searchValue } = useContext(SearchContext)
    
    return searchValue.length === 0;
  }
  
  if (!isCategorisedList()) {
    return (null)
  }
    
  const categorizedContacts = categorizeContacts(contacts, t('empty-list'))

  const letterShortcuts = Object
    .keys(categorizedContacts)
    .filter(letter => letter !== 'EMPTY')
    .sort();

  return (
    <div className="contacts-list-alphabetical-shortcuts">
      {letterShortcuts.map(letter => {
        return <a key={`#cat-${letter}`} href={`#cat-${letter}`}>{letter.toUpperCase()}</a>
      })}
    </div>
  )
}

ContactsListAlphabeticalShortcuts.propTypes = {
  contacts: PropTypes.array.isRequired
}
ContactsListAlphabeticalShortcuts.defaultProps = {}

export default ContactsListAlphabeticalShortcuts
