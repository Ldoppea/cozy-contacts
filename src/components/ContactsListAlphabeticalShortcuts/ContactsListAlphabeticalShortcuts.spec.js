import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'

import AppLike from '../../tests/Applike'
import ContactsListAlphabeticalShortcuts from './ContactsListAlphabeticalShortcuts'

describe('ContactsListAlphabeticalShortcuts', () => {
  it('should display alphabetical shortcuts for 1 letter', () => {
    const contacts = [
        { name: { familyName: 'Doe', givenName: 'John' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "doejohn" } }
    ]
    const alphabeticalShortcutsInstance = (
      <AppLike>
        <ContactsListAlphabeticalShortcuts contacts={contacts} />
      </AppLike>
    )
    const alphabeticalShortcuts = mount(alphabeticalShortcutsInstance)
    const shortcutLinks = alphabeticalShortcuts.find('a')
    expect(shortcutLinks).toBeDefined()
    expect(shortcutLinks.text()).toBe('D')
  })

  it('should display alphabetical shortcuts for 2 letters', () => {
    const contacts = [
        { name: { familyName: 'Backer', givenName: 'Miles' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "backermiles" } },
        { name: { familyName: 'Doe', givenName: 'John' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "doejohn" } },
    ]
    const alphabeticalShortcutsInstance = (
      <AppLike>
        <ContactsListAlphabeticalShortcuts contacts={contacts} />
      </AppLike>
    )
    const alphabeticalShortcuts = mount(alphabeticalShortcutsInstance)
    const shortcutLinks = alphabeticalShortcuts.find('a')
    expect(shortcutLinks).toBeDefined()
    expect(shortcutLinks.length).toBe(2)
    expect(shortcutLinks.at(0).text()).toBe('B')
    expect(shortcutLinks.at(1).text()).toBe('D')
  })

  it('should display alphabetical shortcuts for 2 ordered letters even if contacts list is not ordered', () => {
    const contacts = [
        { name: { familyName: 'Doe', givenName: 'John' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "doejohn" } },
        { name: { familyName: 'Backer', givenName: 'Miles' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "backermiles" } },
    ]
    const alphabeticalShortcutsInstance = (
      <AppLike>
        <ContactsListAlphabeticalShortcuts contacts={contacts} />
      </AppLike>
    )
    const alphabeticalShortcuts = mount(alphabeticalShortcutsInstance)
    const shortcutLinks = alphabeticalShortcuts.find('a')
    expect(shortcutLinks).toBeDefined()
    expect(shortcutLinks.length).toBe(2)
    expect(shortcutLinks.at(0).text()).toBe('B')
    expect(shortcutLinks.at(1).text()).toBe('D')
  })

  it('should accept empty array', () => {
    const contacts = []
    const alphabeticalShortcutsInstance = (
      <AppLike>
        <ContactsListAlphabeticalShortcuts contacts={contacts} />
      </AppLike>
    )
    const alphabeticalShortcuts = mount(alphabeticalShortcutsInstance)
    const shortcutLinks = alphabeticalShortcuts.find('a')
    expect(shortcutLinks).toBeDefined()
    expect(shortcutLinks.length).toBe(0)
  })

  it('should match the alphabetical shortcuts snapshot', () => {
    const contacts = [
        { name: { familyName: 'Backer', givenName: 'Miles' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "backermiles" } },
        { name: { familyName: 'Doe', givenName: 'John' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "doejohn" } },
        { name: { familyName: 'Carroll', givenName: 'Aiden' }, indexes: { byFamilyNameGivenNameEmailCozyUrl: "carrollaiden" } },
    ]

    const tree = renderer
      .create(
        <AppLike>
          <ContactsListAlphabeticalShortcuts contacts={contacts} />
        </AppLike>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
