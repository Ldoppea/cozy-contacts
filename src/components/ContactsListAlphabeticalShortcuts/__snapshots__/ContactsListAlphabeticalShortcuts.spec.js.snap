// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactsListAlphabeticalShortcuts should match the alphabetical shortcuts snapshot 1`] = `
<div
  className="contacts-list-alphabetical-shortcuts"
>
  <a
    href="#cat-b"
  >
    B
  </a>
  <a
    href="#cat-c"
  >
    C
  </a>
  <a
    href="#cat-d"
  >
    D
  </a>
</div>
`;
